core/state.go

```diff
@@ -362,39 +352,97 @@ func (s *State) Class(classHash *felt.Felt) (*DeclaredClass, error) {
-	blockNumber uint64,
-) error {
-	contract, err := NewContract(addr, s.txn)
+func (s *State) updateStorageBuffered(contractAddr *felt.Felt, updateDiff []StorageDiff, blockNumber uint64, logChanges bool) (
+	*db.BufferedTransaction, error,
+) {
+	// to avoid multiple transactions writing to s.txn, create a buffered transaction and use that in the worker goroutine
+	bufferedTxn := db.NewBufferedTransaction(s.txn)
+	bufferedState := NewState(bufferedTxn)
+	bufferedContract, err := NewContract(contractAddr, bufferedTxn)
 	if err != nil {
-		if !errors.Is(err, ErrContractNotDeployed) {
-			return err
+		return nil, err
+	}
+
+	onValueChanged := func(location, oldValue *felt.Felt) error {
+		if logChanges {
+			return bufferedState.LogContractStorage(contractAddr, location, oldValue, blockNumber)
 		}
+		return nil
+	}
 
-		if _, ok := noClassContracts[*addr]; !ok {
-			return err
+	if err = bufferedContract.UpdateStorage(updateDiff, onValueChanged); err != nil {
+		return nil, err
+	}
+
+	return bufferedTxn, nil
+}
+
+// updateContractStorage applies the diff set to the Trie of the
+// contract at the given address in the given Txn context.
+func (s *State) updateContractStorages(stateTrie *trie.Trie, diffs map[felt.Felt][]StorageDiff, blockNumber uint64, logChanges bool) error {
+	// make sure all noClassContracts are deployed
+	for addr := range diffs {
+		if _, ok := noClassContracts[addr]; !ok {
+			continue
 		}
 
-		// Deploy noClassContract
-		err = s.putNewContract(stateTrie, addr, noClassContractsClassHash, blockNumber)
+		_, err := NewContract(&addr, s.txn)
 		if err != nil {
+			if !errors.Is(err, ErrContractNotDeployed) {
+				return err
+			}
+			// Deploy noClassContract
+			err = s.putNewContract(stateTrie, &addr, noClassContractsClassHash, blockNumber)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// sort the contracts in decending diff size order
+	// so we start with the heaviest update first
+	keys := make([]felt.Felt, 0, len(diffs))
+	for key := range diffs {
+		keys = append(keys, key)
+	}
+	sort.SliceStable(keys, func(i, j int) bool {
+		return len(diffs[keys[i]]) > len(diffs[keys[j]])
+	})
+
+	// update per-contract storage Tries concurrently
+	contractUpdaters := pool.NewWithResults[*db.BufferedTransaction]().WithErrors().WithMaxGoroutines(runtime.GOMAXPROCS(0))
+	for _, key := range keys {
+		conractAddr := key
+		updateDiff := diffs[conractAddr]
+		contractUpdaters.Go(func() (*db.BufferedTransaction, error) {
+			return s.updateStorageBuffered(&conractAddr, updateDiff, blockNumber, logChanges)
+		})
+	}
+
+	bufferedTxns, err := contractUpdaters.Wait()
+	if err != nil {
+		return err
+	}
+
+	// flush buffered txns
+	for _, bufferedTxn := range bufferedTxns {
+		if err = bufferedTxn.Flush(); err != nil {
 			return err
 		}
+	}
 
-		// We need to set the contract to the newly deployed noClassContract.
-		contract, err = NewContract(addr, s.txn)
+	for addr := range diffs {
```

Comment:
Nope, because we shouldnt modify `s.txn` before we make sure that all contract storages are updated (ie `contractUpdaters.Wait()` returns). `s.txn` is essentially a `pebble.Batch` and it is not safe to be modified concurrently or modified while another routine is reading from it.

Think of it as a resource that needs to be protected with a `RWMutex`. But by delaying any writes to the transaction until we are done with updating contract storages by using a `BufferedTransaction`, we avoid needing a mutex.

Reply:
Can we continue making the updates in the buffered transaction?

Reply:
During the concurrent contract storage updates?

Yes, because it never writes to the transaction until `Flush()` or `Commit()` is called explicitly. It just keeps track of what changes were made on the `BufferedTransaction` in memory. Essentially it allows our goroutines to access `s.txn` in read-only mode, without them knowing it.

Later on, we apply those changes to `s.txn` all at the same time.

Reply:
https://github.com/NethermindEth/juno/pull/900/files#diff-c075eb780b39817edd44ca2456cee339e3c84f432658a27a58cbb44bc9204780R428-R432

Actually if we sort the slice of `BufferedTransaction`s we get back from the routines in ascending contract address order, we might get an additional speed up as well. But it is not really worth it at this point. Will leave it to another PR.

Reply:
at L435?

No, we can't. because `updateContractCommitment` operates on the state trie regardless of which contract it was called with. All operations on a single Trie instance need to be done sequentially

Reply:
One more related question: why not just use an `RWMutex` instead of creating a new transaction implementation?

Reply:
I expected a lock to have a worse performance impact compared to memory allocations that we need for buffering.

But that is just my gut feeling. We can keep experimenting after merging this.

But regardless, we will have to implement a new type that wraps a transaction and protects it with a RWMutex.

Reply:
Do you mind creating issues for this Mutex thing and the sorting idea? I can if you don't have time.